<a href="https://colab.research.google.com/github/shreyasc84/Prompt-Engineering-Notes/blob/main/m1_assignment_stock_price_prediction_with_llm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800474  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857315  114.354702  111.397715  113.037711  7.636470e+07
50%     133.185188  134.319768  131.222483  133.006499  9.493580e+07
75%     148.415882  149.584513  146.468716  147.782312  1.296327e+08
max     178.879913  179.793905  176.039592  179.489238  4.265100e+08


In [ ]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())



YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed


Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800473  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857307  114.354706  111.397708  113.037705  7.636470e+07
50%     133.185196  134.319791  131.222513  133.006484  9.493580e+07
75%     148.415874  149.584494  146.468701  147.782328  1.296327e+08
max     178.879898  179.793936  176.039622  179.489269  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

In [2]:
import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Derive historical prices as features
data['Open_Lag1'] = data['Open'].shift(1)
data['High_Lag1'] = data['High'].shift(1)
data['Low_Lag1'] = data['Low'].shift(1)
data['Close_Lag1'] = data['Close'].shift(1)
data['Volume_Lag1'] = data['Volume'].shift(1)

# Derive technical indicators (Moving Averages)
data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Drop rows with NaN values created by shifting and rolling windows
data.dropna(inplace=True)

# Display the first few rows with new features
display(data.head())

/tmp/ipython-input-4281943939.py:5: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed


Price           Close       High        Low       Open     Volume  Open_Lag1  \
Ticker           AAPL       AAPL       AAPL       AAPL       AAPL              
Date                                                                           
2020-03-13  67.292786  67.764857  61.235780  64.126297  370732000  61.959609   
2020-03-16  58.635761  62.719755  58.100749  58.572816  322423600  64.126297   
2020-03-17  61.213993  62.363900  57.713421  59.918829  324056000  58.572816   
2020-03-18  59.715466  60.521614  57.403539  58.045070  300233600  59.918829   
2020-03-19  59.257927  61.209143  58.732600  59.889773  271857200  58.045070   

Price       High_Lag1   Low_Lag1 Close_Lag1  Volume_Lag1       MA_5  \
Ticker                                                                
Date                                                                  
2020-03-13  65.363344  60.037442  60.093121  418474000.0  65.515377   
2020-03-16  67.764857  61.235780  67.292786  370732000.0  64.355299   
2020-03-17  62.719755  58.100749  58.635761  322423600.0  62.782706   
2020-03-18  62.363900  57.713421  61.213993  324056000.0  61.390225   
2020-03-19  60.521614  57.403539  59.715466  300233600.0  61.223186   

Price           MA_10      MA_50  
Ticker                            
Date                              
2020-03-13  68.412664  73.941989  
2020-03-16  67.042456  73.663934  
2020-03-17  66.159811  73.451548  
2020-03-18  64.802431  73.197744  
2020-03-19  63.637028  72.941600

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Define features (X) and target (y)
# We will use the engineered features to predict the 'Close' price for the next day
X = data[['Open_Lag1', 'High_Lag1', 'Low_Lag1', 'Close_Lag1', 'Volume_Lag1', 'MA_5', 'MA_10', 'MA_50']]
y = data['Close']

# Shift the target variable to predict the next day's closing price
y = y.shift(-1)

# Drop the last row in X and y as the last y value will be NaN after shifting
X = X.iloc[:-1]
y = y.iloc[:-1]

# Split data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=False) # shuffle=False for time series data

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R²): {r2:.2f}")

Mean Squared Error (MSE): 17.74
R-squared (R²): 0.85


In [4]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import numpy as np

# Create the target variable for classification
# 1 if the close price increases tomorrow, 0 otherwise
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Define features (X) and target (y)
# Use the same features as the regression model
X = data[['Open_Lag1', 'High_Lag1', 'Low_Lag1', 'Close_Lag1', 'Volume_Lag1', 'MA_5', 'MA_10', 'MA_50']]
y = data['Price_Increase']

# Drop the last row in X and y as the last y value will be NaN after shifting
X = X.iloc[:-1]
y = y.iloc[:-1]

# Split data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=False) # shuffle=False for time series data

# Initialize and train the Random Forest Classifier model
model = RandomForestClassifier(n_estimators=100, random_state=42) # Using 100 estimators and a random state for reproducibility
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy:.2f}")
print("Confusion Matrix:")
print(conf_matrix)
print("Classification Report:")
print(class_report)

# Report feature importance
feature_importances = pd.Series(model.feature_importances_, index=X.columns).sort_values(ascending=False)
print("\nFeature Importance:")
print(feature_importances)

Accuracy: 0.51
Confusion Matrix:
[[26 49]
 [20 47]]
Classification Report:
              precision    recall  f1-score   support

           0       0.57      0.35      0.43        75
           1       0.49      0.70      0.58        67

    accuracy                           0.51       142
   macro avg       0.53      0.52      0.50       142
weighted avg       0.53      0.51      0.50       142


Feature Importance:
Price        Ticker
Volume_Lag1            0.133534
Close_Lag1             0.131208
Low_Lag1               0.128840
High_Lag1              0.127618
MA_50                  0.124223
MA_5                   0.124081
Open_Lag1              0.117037
MA_10                  0.113459
dtype: float64


In [5]:
import yfinance as yf
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import numpy as np

# Download historical data for AAPL from 2024-01-01 to 2025-01-01
future_data = yf.download("AAPL", start="2024-01-01", end="2025-01-01")

# Apply the same feature engineering steps to the new data
future_data['Open_Lag1'] = future_data['Open'].shift(1)
future_data['High_Lag1'] = future_data['High'].shift(1)
future_data['Low_Lag1'] = future_data['Low'].shift(1)
future_data['Close_Lag1'] = future_data['Close'].shift(1)
future_data['Volume_Lag1'] = future_data['Volume'].shift(1)
future_data['MA_5'] = future_data['Close'].rolling(window=5).mean()
future_data['MA_10'] = future_data['Close'].rolling(window=10).mean()
future_data['MA_50'] = future_data['Close'].rolling(window=50).mean()

# Create the target variable for classification in the new data
future_data['Price_Increase'] = (future_data['Close'].shift(-1) > future_data['Close']).astype(int)

# Drop rows with NaN values created by shifting and rolling windows
future_data.dropna(inplace=True)

# Define features (X_future) and target (y_future) for the future data
X_future = future_data[['Open_Lag1', 'High_Lag1', 'Low_Lag1', 'Close_Lag1', 'Volume_Lag1', 'MA_5', 'MA_10', 'MA_50']]
y_future = future_data['Price_Increase']

# Drop the last row in X_future and y_future as the last y_future value will be NaN after shifting
X_future = X_future.iloc[:-1]
y_future = y_future.iloc[:-1]

# Ensure the columns in X_future match the columns used for training
# This is important if any columns were dropped or added during feature engineering
# In this case, the columns should be the same, but it's a good practice to ensure
X_future = X_future[X_train.columns]


# Make predictions on the future data using the trained Random Forest model
y_pred_future = model.predict(X_future)

# Evaluate the model performance on the future data
accuracy_future = accuracy_score(y_future, y_pred_future)
conf_matrix_future = confusion_matrix(y_future, y_pred_future)
class_report_future = classification_report(y_future, y_pred_future)

print(f"Accuracy on 2024-2025 data: {accuracy_future:.2f}")
print("Confusion Matrix on 2024-2025 data:")
print(conf_matrix_future)
print("Classification Report on 2024-2025 data:")
print(class_report_future)

/tmp/ipython-input-1723417692.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  future_data = yf.download("AAPL", start="2024-01-01", end="2025-01-01")
[*********************100%***********************]  1 of 1 completed

Accuracy on 2024-2025 data: 0.39
Confusion Matrix on 2024-2025 data:
[[ 78   0]
 [123   1]]
Classification Report on 2024-2025 data:
              precision    recall  f1-score   support

           0       0.39      1.00      0.56        78
           1       1.00      0.01      0.02       124

    accuracy                           0.39       202
   macro avg       0.69      0.50      0.29       202
weighted avg       0.76      0.39      0.23       202



# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.

